tower-filter/src/error.rs

```diff
@@ -0,0 +1,54 @@
+//! Error types
+
+use std::error;
+use std::fmt;
+
+/// Error produced by `Filter`
+#[derive(Debug)]
+pub struct Error {
+    source: Option<Source>,
+}
+
+pub(crate) type Source = Box<dyn error::Error + Send + Sync>;
+
+impl Error {
+    /// Create a new `Error` representing a rejected request.
+    pub fn rejected() -> Error {
+        Error { source: None }
+    }
+
+    /// Create a new `Error` representing an inner service error.
+    pub fn inner<E>(source: E) -> Error
+    where
+        E: Into<Source>,
+    {
+        Error {
+            source: Some(source.into()),
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        if let Some(source) = &self.source {
+            source.fmt(fmt)
+        } else {
+            write!(fmt, "rejected")
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        if let Some(ref err) = self.source {
```

Comment:
Either this shouldn't return the inner error, or the `Display` shouldn't output the inner error (only). Otherwise, something that iterates the error source chain and outputs sees this:

```
Error: my inner error
Caused by: my inner error
```

Reply:
@seanmonstar probably should be `Display` that is fixed then?